docker/puppet-agent-alpine/Dockerfile

```diff
@@ -9,97 +9,85 @@ SHELL ["/bin/ash", "-eo", "pipefail", "-c"]
 
 RUN mkdir /workspace
 WORKDIR /workspace
-RUN sed -i -e 's/sys\/poll/poll/' /usr/include/boost/asio/detail/socket_types.hpp
 
-COPY configs/components/leatherman.json /workspace
-RUN git clone -b 1.5.x https://github.com/puppetlabs/leatherman && \
-    mkdir -p /workspace/leatherman/build
+COPY configs/components/leatherman.json \
+     configs/components/libwhereami.json \
+     configs/components/cpp-hocon.json \
+     configs/components/facter.json \
+     configs/components/cpp-pcp-client.json \
+     configs/components/pxp-agent.json \
+     configs/components/hiera.json \
+     configs/components/puppet.json \
+     /workspace/
+
+RUN sed -i -e 's/sys\/poll/poll/' /usr/include/boost/asio/detail/socket_types.hpp && \
+    git clone -b 1.5.x https://github.com/puppetlabs/leatherman && \
+    mkdir -p /workspace/leatherman/build && \
+    git clone https://github.com/puppetlabs/libwhereami && \
+    mkdir -p /workspace/libwhereami/build && \
+    git clone https://github.com/puppetlabs/cpp-hocon && \
+    mkdir -p /workspace/cpp-hocon/build && \
+    git clone https://github.com/puppetlabs/cpp-pcp-client && \
+    mkdir -p /workspace/cpp-pcp-client/build && \
+    git clone https://github.com/puppetlabs/facter && \
+    mkdir -p /workspace/facter/build && \
+    git clone https://github.com/puppetlabs/pxp-agent && \
+    mkdir -p /workspace/pxp-agent/build && \
+    curl -O -L https://people.redhat.com/~rjones/virt-what/files/virt-what-1.18.tar.gz && \
+    tar zxf virt-what-1.18.tar.gz && \
+    git clone https://github.com/puppetlabs/hiera && \
+    git clone https://github.com/puppetlabs/puppet
+
```

Comment:
so, these were all split up by project before to be able to make better use of reusing layers so that you don't have to rebuild everything if just, say, puppet changes

Reply:
If I'm not mistaken, any `RUN` command immediately invalidates the Docker build cache from that point forward... so I don't think Docker is smart enough to understand what you were aiming for.

But I could be mistaken - so let me re-verify my understanding since this is one of the more complex containers given its multi-stage.

Reply:
Ok, for existing build:

* first stage is able to cache 7/12 steps
* second stage is able to cache 71/81 steps 

For updated: